### PR TITLE
[move][easy] Add json error support to move compilation

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

Add JSON error reporting support to move compilation (closes https://github.com/MystenLabs/sui/issues/15755)

## Test plan 

New test added

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs [hunk 1]
```diff
@@ -59,7 +59,7 @@ pub struct Diagnostic {
 }
 
 #[derive(PartialEq, Eq, Hash, Clone, Debug, Default)]
-pub struct Diagnostics(Option<Diagnostics_>);
+pub struct Diagnostics(Option<Diagnostics_>, DiagnosticsFormat);
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:62]: nit, maybe named fields now?

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs [hunk 2]
```diff
@@ -361,6 +379,62 @@ fn render_diagnostic(
...
+
+    writeln!(&mut output, "    {{").expect("ICE reporting error");
+    let bloc = mapped_files.location(ploc);
+    writeln!(
+        &mut output,
+        "        \"file\": \"{}\",",
+        mapped_files.files.get(bloc.file_id).unwrap().name()
+    )
+    .expect("ICE reporting error");
+    writeln!(&mut output, "        \"line\": {},", bloc.start.line).expect("ICE reporting error");
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:425]: not necessary, but should we just spit this out into some struct and rely on serde_json?

**Review by tnowacki — APPROVED** [review]

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs (hunk 2 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:425]: Excellent idea! Done.

File: external-crates/move/crates/move-compiler/src/diagnostics/mod.rs (hunk 1 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/diagnostics/mod.rs:62]: I went back and forth on this, so if you think so, too, I'll do it.

**Pull request merged**